{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [-l SLIBPATH] [-m NMATCH] [-x NMISMATCH]\n",
      "                             [-o NOPEN] [-e NEXT] [-p] [-a SMATRIX] [-c]\n",
      "                             [-f NTHR] [-r] [-s] [-header]\n",
      "                             target query\n",
      "ipykernel_launcher.py: error: the following arguments are required: target, query\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/asad/anaconda3/envs/deeplearn/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3339: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "\"\"\"\n",
    "Simple python wrapper for SSW library\n",
    "Please put the path of libssw.so into LD_LIBRARY_PATH or pass it explicitly as a parameter\n",
    "By Yongan Zhao (March 2016)\n",
    "\"\"\"\n",
    "\n",
    "import sys\n",
    "import os.path as op\n",
    "import argparse as ap\n",
    "import ctypes as ct\n",
    "import timeit as ti\n",
    "import gzip\n",
    "import math\n",
    "\n",
    "import ssw_lib\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def read(sFile):\n",
    "    \"\"\"\n",
    "    read a sequence file\n",
    "    @param  sFile   sequence file\n",
    "    \"\"\"\n",
    "    def read_one_fasta(f):\n",
    "        \"\"\"\n",
    "        read a fasta file\n",
    "        @param  f   file handler\n",
    "        \"\"\"\n",
    "        sId = ''\n",
    "        sSeq = ''\n",
    "        for l in f:\n",
    "            if l.startswith('>'):\n",
    "                if sSeq:\n",
    "                    yield sId, sSeq, ''\n",
    "                sId = l.strip()[1:].split()[0]\n",
    "                sSeq = ''\n",
    "            else:\n",
    "                sSeq += l.strip()\n",
    "\n",
    "        yield sId, sSeq, ''\n",
    "\n",
    "    def read_one_fastq(f):\n",
    "        \"\"\"\n",
    "        read a fastq file\n",
    "        @param  f   file handler\n",
    "        \"\"\"\n",
    "        sId = ''\n",
    "        sSeq = ''\n",
    "        s3 = ''\n",
    "        sQual = ''\n",
    "        for l in f:\n",
    "            sId = l.strip()[1:].split()[0]\n",
    "            sSeq = f.next()\n",
    "            s3 = f.next()\n",
    "            sQual = f.next()\n",
    "\n",
    "            yield sId, sSeq, sQual\n",
    "\n",
    "# test if fasta or fastq\n",
    "    bFasta = True\n",
    "    ext = op.splitext(sFile)[1][1:].strip().lower()\n",
    "    if ext == 'gz' or ext == 'gzip':\n",
    "        with gzip.open(sFile, 'r') as f:\n",
    "            l = f.next()\n",
    "            if l.startswith('>'):\n",
    "                bFasta = True\n",
    "            elif l.startswith('@'):\n",
    "                bFasta = False\n",
    "            else:\n",
    "                print >> sys.stderr, 'file format cannot be recognized'\n",
    "                sys.exit()\n",
    "    else:\n",
    "        with open(sFile, 'r') as f:\n",
    "            l = f.next()\n",
    "            if l.startswith('>'):\n",
    "                bFasta = True\n",
    "            elif l.startswith('@'):\n",
    "                bFasta = False\n",
    "            else:\n",
    "                print >> sys.stderr, 'file format cannot be recognized'\n",
    "                sys.exit()\n",
    "\n",
    "# read\n",
    "    if ext == 'gz' or ext == 'gzip':\n",
    "        with gzip.open(sFile, 'r') as f:\n",
    "            if bFasta == True:\n",
    "                for sId,sSeq,sQual in read_one_fasta(f):\n",
    "                    yield sId, sSeq, sQual\n",
    "            else:\n",
    "                for sId,sSeq,sQual in read_one_fastq(f):\n",
    "                    yield sId, sSeq, sQual\n",
    "    else:\n",
    "        with open(sFile, 'r') as f:\n",
    "            if bFasta == True:\n",
    "                for sId,sSeq,sQual in read_one_fasta(f):\n",
    "                    yield sId, sSeq, sQual\n",
    "            else:\n",
    "                for sId,sSeq,sQual in read_one_fastq(f):\n",
    "                    yield sId, sSeq, sQual\n",
    "\n",
    "\n",
    "def to_int(seq, lEle, dEle2Int):\n",
    "    \"\"\"\n",
    "    translate a sequence into numbers\n",
    "    @param  seq   a sequence\n",
    "    \"\"\"\n",
    "    num_decl = len(seq) * ct.c_int8\n",
    "    num = num_decl()\n",
    "    for i,ele in enumerate(seq):\n",
    "        try:\n",
    "            n = dEle2Int[ele]\n",
    "        except KeyError:\n",
    "            n = dEle2Int[lEle[-1]]\n",
    "        finally:\n",
    "            num[i] = n\n",
    "\n",
    "    return num\n",
    "\n",
    "\n",
    "def align_one(ssw, qProfile, rNum, nRLen, nOpen, nExt, nFlag, nMaskLen):\n",
    "    \"\"\"\n",
    "    align one pair of sequences\n",
    "    @param  qProfile   query profile\n",
    "    @param  rNum   number array for reference\n",
    "    @param  nRLen   length of reference sequence\n",
    "    @param  nFlag   alignment flag\n",
    "    @param  nMaskLen   mask length\n",
    "    \"\"\"\n",
    "    res = ssw.ssw_align(qProfile, rNum, ct.c_int32(nRLen), nOpen, nExt, nFlag, 0, 0, nMaskLen)\n",
    "\n",
    "    nScore = res.contents.nScore\n",
    "    nScore2 = res.contents.nScore2\n",
    "    nRefBeg = res.contents.nRefBeg\n",
    "    nRefEnd = res.contents.nRefEnd\n",
    "    nQryBeg = res.contents.nQryBeg\n",
    "    nQryEnd = res.contents.nQryEnd\n",
    "    nRefEnd2 = res.contents.nRefEnd2\n",
    "    lCigar = [res.contents.sCigar[idx] for idx in range(res.contents.nCigarLen)]\n",
    "    nCigarLen = res.contents.nCigarLen\n",
    "    ssw.align_destroy(res)\n",
    "\n",
    "    return (nScore, nScore2, nRefBeg, nRefEnd, nQryBeg, nQryEnd, nRefEnd2, nCigarLen, lCigar)\n",
    "\n",
    "\n",
    "def buildPath(q, r, nQryBeg, nRefBeg, lCigar):\n",
    "    \"\"\"\n",
    "    build cigar string and align path based on cigar array returned by ssw_align\n",
    "    @param  q   query sequence\n",
    "    @param  r   reference sequence\n",
    "    @param  nQryBeg   begin position of query sequence\n",
    "    @param  nRefBeg   begin position of reference sequence\n",
    "    @param  lCigar   cigar array\n",
    "    \"\"\"\n",
    "    sCigarInfo = 'MIDNSHP=X'\n",
    "    sCigar = ''\n",
    "    sQ = ''\n",
    "    sA = ''\n",
    "    sR = ''\n",
    "    nQOff = nQryBeg\n",
    "    nROff = nRefBeg\n",
    "    for x in lCigar:\n",
    "        n = x >> 4\n",
    "        m = x & 15\n",
    "        if m > 8:\n",
    "            c = 'M'\n",
    "        else:\n",
    "            c = sCigarInfo[m]\n",
    "        sCigar += str(n) + c\n",
    "\n",
    "        if c == 'M':\n",
    "            sQ += q[nQOff : nQOff+n]\n",
    "            sA += ''.join(['|' if q[nQOff+j] == r[nROff+j] else '*' for j in xrange(n)])\n",
    "            sR += r[nROff : nROff+n]\n",
    "            nQOff += n\n",
    "            nROff += n\n",
    "        elif c == 'I':\n",
    "            sQ += q[nQOff : nQOff+n]\n",
    "            sA += ' ' * n\n",
    "            sR += '-' * n\n",
    "            nQOff += n\n",
    "        elif c == 'D':\n",
    "            sQ += '-' * n\n",
    "            sA += ' ' * n\n",
    "            sR += r[nROff : nROff+n]\n",
    "            nROff += n\n",
    "\n",
    "    return sCigar, sQ, sA, sR\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def main(args):\n",
    "    lEle = []\n",
    "    dRc = {} \n",
    "    dEle2Int = {}\n",
    "    dInt2Ele = {}\n",
    "    if False == args.bProtien:\n",
    "# init DNA score matrix\n",
    "        if not args.sMatrix:\n",
    "            lEle = ['A', 'C', 'G', 'T', 'N']\n",
    "            dRc = {'A':'C', 'C':'G', 'G':'C', 'T':'A', 'a':'C', 'c':'G', 'g':'C', 't':'A'} \n",
    "            for i,ele in enumerate(lEle):\n",
    "                dEle2Int[ele] = i\n",
    "                dEle2Int[ele.lower()] = i\n",
    "                dInt2Ele[i] = ele\n",
    "            nEleNum = len(lEle)\n",
    "            lScore = [0 for i in xrange(nEleNum**2)]\n",
    "            for i in xrange(nEleNum-1):\n",
    "                for j in xrange(nEleNum-1):\n",
    "                    if lEle[i] == lEle[j]:\n",
    "                        lScore[i*nEleNum+j] = args.nMatch\n",
    "                    else:\n",
    "                        lScore[i*nEleNum+j] = -args.nMismatch\n",
    "        else:\n",
    "            lEle, dEle2Int, dInt2Ele, lScore = ssw.read_matrix(args.sMatrix)\n",
    "    else:\n",
    "# load AA score matrix\n",
    "        if not args.sMatrix:\n",
    "            lEle = 'A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V   B   Z   X   *'.split()\n",
    "            for i,ele in enumerate(lEle):\n",
    "                dEle2Int[ele] = i\n",
    "                dEle2Int[ele.lower()] = i\n",
    "                dInt2Ele[i] = ele\n",
    "            nEleNum = len(lEle)\n",
    "            lScore = ssw_lib.lBlosum50\n",
    "        else:\n",
    "            # assume the format of the input score matrix is the same as that of http://www.ncbi.nlm.nih.gov/Class/FieldGuide/BLOSUM62.txt\n",
    "            lEle, dEle2Int, dInt2Ele, lScore = ssw.read_matrix(args.sMatrix)\n",
    "\n",
    "    if args.bBest and args.bProtien:\n",
    "        print >> sys.stderr, 'Reverse complement alignment is not available for protein sequences.'\n",
    "\n",
    "# translate score matrix to ctypes\n",
    "    mat = (len(lScore) * ct.c_int8) ()\n",
    "    mat[:] = lScore\n",
    "# set flag\n",
    "    nFlag = 0\n",
    "    if args.bPath:\n",
    "        nFlag = 2\n",
    "# print sam head\n",
    "    if args.bSam and args.bHeader and args.bPath:\n",
    "        print ('@HD\\tVN:1.4\\tSO:queryname')\n",
    "        for sRId,sRSeq,_ in read(args.target):\n",
    "            print ('@SQ\\tSN:{}\\tLN:{}'.format(sRId, len(sRSeq)))\n",
    "    elif args.bSam and not args.bPath:\n",
    "        print >> sys.stderr, 'SAM format output is only available together with option -c.\\n'\n",
    "        args.bSam = False\n",
    "\n",
    "    ssw = ssw_lib.CSsw(args.sLibPath)\n",
    "# iterate query sequence\n",
    "    for sQId,sQSeq,sQQual in read(args.query):\n",
    "# build query profile\n",
    "        qNum = to_int(sQSeq, lEle, dEle2Int)\n",
    "        qProfile = ssw.ssw_init(qNum, ct.c_int32(len(sQSeq)), mat, len(lEle), 2)\n",
    "# build rc query profile\n",
    "        if args.bBest and not args.bProtien:\n",
    "            sQRcSeq = ''.join([dRc[x] for x in sQSeq[::-1]])\n",
    "            qRcNum = to_int(sQRcSeq, lEle, dEle2Int)\n",
    "            qRcProfile = ssw.ssw_init(qRcNum, ct.c_int32(len(sQSeq)), mat, len(lEle), 2)\n",
    "# set mask len\n",
    "        if len(sQSeq) > 30:\n",
    "            nMaskLen = len(sQSeq) / 2\n",
    "        else:\n",
    "            nMaskLen = 15\n",
    "\n",
    "# iter target sequence\n",
    "        for sRId,sRSeq,_ in read(args.target):\n",
    "            rNum = to_int(sRSeq, lEle, dEle2Int)\n",
    "\n",
    "# format ofres: (nScore, nScore2, nRefBeg, nRefEnd, nQryBeg, nQryEnd, nRefEnd2, nCigarLen, lCigar)\n",
    "            res = align_one(ssw, qProfile, rNum, len(sRSeq), args.nOpen, args.nExt, nFlag, nMaskLen)\n",
    "# align rc query\n",
    "            resRc = None\n",
    "            if args.bBest and not args.bProtien:\n",
    "                resRc = align_one(ssw, qRcProfile, rNum, len(sRSeq), args.nOpen, args.nExt, nFlag, nMaskLen)\n",
    "\n",
    "# build cigar and trace back path\n",
    "            strand = 0\n",
    "            if resRc == None or res[0] > resRc[0]:\n",
    "                resPrint = res\n",
    "                strand = 0\n",
    "                sCigar, sQ, sA, sR = buildPath(sQSeq, sRSeq, res[4], res[2], res[8])\n",
    "            else:\n",
    "                resPrint = resRc\n",
    "                strand = 1\n",
    "                sCigar, sQ, sA, sR = buildPath(sQRcSeq, sRSeq, resRc[4], resRc[2], resRc[8])\n",
    "\n",
    "# print results\n",
    "            if not args.bSam:\n",
    "                print ('target_name: {}\\nquery_name: {}\\noptimal_alignment_score: {}\\t'.format(sRId, sQId, resPrint[0])),\n",
    "                if resPrint[1] > 0:\n",
    "                    print ('suboptimal_alignment_score: {}\\t'.format(resPrint[1])),\n",
    "                if strand == 0:\n",
    "                    print ('strand: +\\t'),\n",
    "                else: \n",
    "                    print ('strand: -\\t'),\n",
    "                if resPrint[2] + 1:\n",
    "                    print ('target_begin: {}\\t'.format(resPrint[2] + 1)),\n",
    "                print ('target_end: {}\\t'.format(resPrint[3] + 1)),\n",
    "                if resPrint[4] + 1:\n",
    "                    print ('query_begin: {}\\t'.format(resPrint[4] + 1)),\n",
    "                print ('query_end: {}\\n'.format(resPrint[5] + 1))\n",
    "                if resPrint[-2] > 0:\n",
    "                    n1 = 1 + resPrint[2]\n",
    "                    n2 = min(60,len(sR)) + resPrint[2] - sR.count('-',0,60)\n",
    "                    n3 = 1 + resPrint[4]\n",
    "                    n4 = min(60,len(sQ)) + resPrint[4] - sQ.count('-',0,60)\n",
    "                    for i in range(0, len(sQ), 60):\n",
    "                        print ('Target:{:>8}\\t{}\\t{}'.format(n1, sR[i:i+60], n2))\n",
    "                        n1 = n2 + 1\n",
    "                        n2 = n2 + min(60,len(sR)-i-60) - sR.count('-',i+60,i+120)\n",
    "\n",
    "                        print ('{: ^15}\\t{}'.format('', sA[i:i+60]))\n",
    "\n",
    "                        print ('Query:{:>9}\\t{}\\t{}\\n'.format(n3, sQ[i:i+60], n4))\n",
    "                        n3 = n4 + 1\n",
    "                        n4 = n4 + min(60,len(sQ)-i-60) - sQ.count('-',i+60,i+120)\n",
    "            else:\n",
    "                print (\"{}\\t\".format(sQId)),\n",
    "                if resPrint[0] == 0:\n",
    "                    print (\"4\\t*\\t0\\t255\\t*\\t*\\t0\\t0\\t*\\t*\"),\n",
    "                else:\n",
    "                    mapq = int(-4.343 * math.log(1-abs(resPrint[0]-resPrint[1])/float(resPrint[0])))\n",
    "                    mapq = int(mapq + 4.99);\n",
    "                    if mapq >= 254:\n",
    "                        mapq = 254\n",
    "                    if strand == 1:\n",
    "                        print ('16\\t'),\n",
    "                    else:\n",
    "                        print ('0\\t'),\n",
    "                    print ('{}\\t{}\\t{}\\t'.format(sRId, resPrint[2]+1, mapq)),\n",
    "                    print (sCigar),\n",
    "                    print ('\\t*\\t0\\t0\\t'),\n",
    "                    print (sQSeq[resPrint[4]:resPrint[5]+1]) if strand==0 else sQRcSeq[resPrint[4]:resPrint[5]+1],\n",
    "                    print ('\\t'),\n",
    "                    if sQQual:\n",
    "                        if strand == 0:\n",
    "                            print (sQQual[resPrint[4]:resPrint[5]+1]),\n",
    "                        else:\n",
    "                            print (sQQual[-resPrint[4]-1:-resPrint[5]-1:-1])\n",
    "                    else:\n",
    "                        print ('*'),\n",
    "\n",
    "                    print ('\\tAS:i:{}'.format(resPrint[0])),\n",
    "                    print ('\\tNM:i:{}\\t'.format(len(sA)-sA.count('|'))),\n",
    "                    if resPrint[1] > 0:\n",
    "                        print ('ZS:i:{}'.format(resPrint[1]))\n",
    "                    else:\n",
    "                        print ('')\n",
    "\n",
    "\n",
    "        ssw.init_destroy(qProfile)\n",
    "        if args.bBest and not args.bProtien:\n",
    "            ssw.init_destroy(qRcProfile)\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    parser = ap.ArgumentParser()\n",
    "    parser.add_argument('-l', '--sLibPath', default='', help='path of libssw.so')\n",
    "    parser.add_argument('-m', '--nMatch', type=int, default=2, help='a positive integer as the score for a match in genome sequence alignment. [default: 2]')\n",
    "    parser.add_argument('-x', '--nMismatch', type=int, default=2, help='a positive integer as the score for a mismatch in genome sequence alignment. [default: 2]')\n",
    "    parser.add_argument('-o', '--nOpen', type=int, default=3, help='a positive integer as the penalty for the gap opening in genome sequence alignment. [default: 3]')\n",
    "    parser.add_argument('-e', '--nExt', type=int, default=1, help='a positive integer as the penalty for the gap extension in genome sequence alignment. [default: 1]')\n",
    "    parser.add_argument('-p', '--bProtien', action='store_true', help='Do protein sequence alignment. Without this option, the ssw_test will do genome sequence alignment. [default: False]')\n",
    "    parser.add_argument('-a', '--sMatrix', default='', help='a file for either Blosum or Pam weight matrix. [default: Blosum50]')\n",
    "    parser.add_argument('-c', '--bPath', action='store_true', help='Return the alignment path. [default: False]')\n",
    "    parser.add_argument('-f', '--nThr', default=0, help='a positive integer. Only output the alignments with the Smith-Waterman score >= N.')\n",
    "    parser.add_argument('-r', '--bBest', action='store_true', help='The best alignment will be picked between the original read alignment and the reverse complement read alignment. [default: False]')\n",
    "    parser.add_argument('-s', '--bSam', action='store_true', help='Output in SAM format. [default: no header]')\n",
    "    parser.add_argument('-header', '--bHeader', action='store_true', help='If -s is used, include header in SAM output.')\n",
    "    parser.add_argument('target', help='targe file')\n",
    "    parser.add_argument('query', help='query file')\n",
    "    if len(sys.argv) == 1:\n",
    "        parser.print_help()\n",
    "        sys.exit()\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    t1 = ti.default_timer()\n",
    "    main(args)\n",
    "    t2 = ti.default_timer()\n",
    "    print >> sys.stderr, 'CPU time: {} seconds'.format(t2 - t1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
